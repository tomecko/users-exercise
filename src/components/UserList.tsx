import React from "react";
import { HiPencil, HiTrash } from "react-icons/hi2";
import { User } from "../model";

interface Props {
  users: User[];
  onEdit: (userId: User["id"], partial: Partial<User>) => void;
  onRemove: (userId: User["id"]) => void;
}

const UserList: React.FC<Props> = ({ users, onEdit, onRemove }) => {
  return (
    <ul>
      {users.map((user) => (
        <li
          key={user.id}
          className="group/user max-w-2xl border relative mb-2 p-2 pr-6 bg-blue-100 border-blue-400"
          data-testid="user"
        >
          {(["name", "email"] as const).map((key) =>
            UserInfo({ propKey: key, user })
          )}
          {RemoveUser({ user })}
        </li>
      ))}
    </ul>
  );

  function UserInfo({
    propKey,
    user,
  }: {
    propKey: "name" | "email";
    user: User;
  }) {
    return (
      <div className="group" key={propKey}>
        {user[propKey]}
        <button
          className="text-gray-600 invisible group-hover:visible"
          data-testid={`edit-user-${propKey}`}
          onClick={() => {
            const newValue =
              prompt(`Enter new ${propKey}`, user[propKey]) ?? "";
            onEdit(user.id, { [propKey]: newValue });
          }}
        >
          <HiPencil className="ml-2" />
        </button>
      </div>
    );
  }

  function RemoveUser({ user }: { user: User }) {
    return (
      <button
        className="absolute top-1/2 -translate-y-1/2 right-2 text-gray-600 invisible group-hover/user:visible"
        data-testid="remove-user"
        onClick={() => {
          onRemove(user.id);
        }}
      >
        <HiTrash />
      </button>
    );
  }
};

export default UserList;
