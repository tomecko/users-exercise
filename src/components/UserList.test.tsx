import React from "react";
import { render } from "@testing-library/react";
import UserList from "./UserList";
import { User } from "../model";

describe("UserList", () => {
  const users: User[] = [
    { id: "1", name: "Alice", email: "alice@example.com" },
    { id: "2", name: "Bob", email: "bob@example.com" },
  ];

  it("renders correctly", () => {
    const { container } = render(
      <UserList users={users} onEdit={() => {}} onRemove={() => {}} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
