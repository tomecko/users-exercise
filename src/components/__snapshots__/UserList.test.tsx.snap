// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList renders correctly 1`] = `
<ul>
  <li
    class="group/user max-w-2xl border relative mb-2 p-2 pr-6 bg-blue-100 border-blue-400"
    data-testid="user"
  >
    <div
      class="group"
    >
      Alice
      <button
        class="text-gray-600 invisible group-hover:visible"
        data-testid="edit-user-name"
      >
        <svg
          aria-hidden="true"
          class="ml-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.731 2.269a2.625 2.625 0 00-3.712 0l-1.157 1.157 3.712 3.712 1.157-1.157a2.625 2.625 0 000-3.712zM19.513 8.199l-3.712-3.712-12.15 12.15a5.25 5.25 0 00-1.32 2.214l-.8 2.685a.75.75 0 00.933.933l2.685-.8a5.25 5.25 0 002.214-1.32L19.513 8.2z"
          />
        </svg>
      </button>
    </div>
    <div
      class="group"
    >
      alice@example.com
      <button
        class="text-gray-600 invisible group-hover:visible"
        data-testid="edit-user-email"
      >
        <svg
          aria-hidden="true"
          class="ml-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.731 2.269a2.625 2.625 0 00-3.712 0l-1.157 1.157 3.712 3.712 1.157-1.157a2.625 2.625 0 000-3.712zM19.513 8.199l-3.712-3.712-12.15 12.15a5.25 5.25 0 00-1.32 2.214l-.8 2.685a.75.75 0 00.933.933l2.685-.8a5.25 5.25 0 002.214-1.32L19.513 8.2z"
          />
        </svg>
      </button>
    </div>
    <button
      class="absolute top-1/2 -translate-y-1/2 right-2 text-gray-600 invisible group-hover/user:visible"
      data-testid="remove-user"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M16.5 4.478v.227a48.816 48.816 0 013.878.512.75.75 0 11-.256 1.478l-.209-.035-1.005 13.07a3 3 0 01-2.991 2.77H8.084a3 3 0 01-2.991-2.77L4.087 6.66l-.209.035a.75.75 0 01-.256-1.478A48.567 48.567 0 017.5 4.705v-.227c0-1.564 1.213-2.9 2.816-2.951a52.662 52.662 0 013.369 0c1.603.051 2.815 1.387 2.815 2.951zm-6.136-1.452a51.196 51.196 0 013.273 0C14.39 3.05 15 3.684 15 4.478v.113a49.488 49.488 0 00-6 0v-.113c0-.794.609-1.428 1.364-1.452zm-.355 5.945a.75.75 0 10-1.5.058l.347 9a.75.75 0 101.499-.058l-.346-9zm5.48.058a.75.75 0 10-1.498-.058l-.347 9a.75.75 0 001.5.058l.345-9z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </li>
  <li
    class="group/user max-w-2xl border relative mb-2 p-2 pr-6 bg-blue-100 border-blue-400"
    data-testid="user"
  >
    <div
      class="group"
    >
      Bob
      <button
        class="text-gray-600 invisible group-hover:visible"
        data-testid="edit-user-name"
      >
        <svg
          aria-hidden="true"
          class="ml-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.731 2.269a2.625 2.625 0 00-3.712 0l-1.157 1.157 3.712 3.712 1.157-1.157a2.625 2.625 0 000-3.712zM19.513 8.199l-3.712-3.712-12.15 12.15a5.25 5.25 0 00-1.32 2.214l-.8 2.685a.75.75 0 00.933.933l2.685-.8a5.25 5.25 0 002.214-1.32L19.513 8.2z"
          />
        </svg>
      </button>
    </div>
    <div
      class="group"
    >
      bob@example.com
      <button
        class="text-gray-600 invisible group-hover:visible"
        data-testid="edit-user-email"
      >
        <svg
          aria-hidden="true"
          class="ml-2"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21.731 2.269a2.625 2.625 0 00-3.712 0l-1.157 1.157 3.712 3.712 1.157-1.157a2.625 2.625 0 000-3.712zM19.513 8.199l-3.712-3.712-12.15 12.15a5.25 5.25 0 00-1.32 2.214l-.8 2.685a.75.75 0 00.933.933l2.685-.8a5.25 5.25 0 002.214-1.32L19.513 8.2z"
          />
        </svg>
      </button>
    </div>
    <button
      class="absolute top-1/2 -translate-y-1/2 right-2 text-gray-600 invisible group-hover/user:visible"
      data-testid="remove-user"
    >
      <svg
        aria-hidden="true"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M16.5 4.478v.227a48.816 48.816 0 013.878.512.75.75 0 11-.256 1.478l-.209-.035-1.005 13.07a3 3 0 01-2.991 2.77H8.084a3 3 0 01-2.991-2.77L4.087 6.66l-.209.035a.75.75 0 01-.256-1.478A48.567 48.567 0 017.5 4.705v-.227c0-1.564 1.213-2.9 2.816-2.951a52.662 52.662 0 013.369 0c1.603.051 2.815 1.387 2.815 2.951zm-6.136-1.452a51.196 51.196 0 013.273 0C14.39 3.05 15 3.684 15 4.478v.113a49.488 49.488 0 00-6 0v-.113c0-.794.609-1.428 1.364-1.452zm-.355 5.945a.75.75 0 10-1.5.058l.347 9a.75.75 0 101.499-.058l-.346-9zm5.48.058a.75.75 0 10-1.498-.058l-.347 9a.75.75 0 001.5.058l.345-9z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </li>
</ul>
`;
